charms/istio-pilot/src/charm.py

```diff
@@ -126,6 +139,58 @@ def remove(self, event):
             manifests, namespace=self.model.name, ignore_not_found=True, ignore_unauthorized=True
         )
 
+    def upgrade_charm(self, event):
+        """Upgrade charm.
+
+        Supports upgrade of exactly one minor version at a time.
+        """
+        istioctl = Istioctl(ISTIOCTL_PATH, self.model.name, ISTIOCTL_DEPOYMENT_PROFILE)
+        self._log_and_set_status(MaintenanceStatus("Upgrading Istio"))
+
+        # Check for version compatability for the upgrade
+        try:
+            versions = istioctl.version()
+        except VersionCheckError as e:
+            self.log.error(UPGRADE_FAILED_MSG.format(message=str(e)))
+            raise GenericCharmRuntimeError("Failed to upgrade.  See `juju debug-log` for details.") from e
```

Comment:
I think this line and the previous one can send a different message. IMO, they should say the version of control plane and client could not be retrieved and because of that the upgrade will not proceed. This check didn't even try to upgrade, so it's fair to have a more specific message of what happened.

Reply:
as is if we fail here, we get a a status of:
> Failed to upgrade.  See `juju debug-log` for details.

and in the debug-log we get:
> Failed to upgrade Istio.  Failed to get client version - no version found in output.  To recover Istio, use istioctl to ...

Would the debug-log message be better as: 
> Failed to upgrade Istio because of an error retrieving version information about istio.  Got message: '{message}' when trying to retrieve version information.

?

I wonder if I should add some general "if an upgrade goes wrong, here's some things to try" instructions to the README.md in this repo.  It is probably easier to have general instructions there, and then it also can be updated live.  Thoughts?

Reply:
For now, I've changed it to the new message listed, but curious what you think

Reply:
Updated things a little more. I've added to the istio-pilot README.md some general instructions on fixing a failed upgrade, and then I've linked the doc in the error message. What do you think?